### DISCUSSION TITLE
Load and Save memory

### DISCUSSION THREAD

**Discussion by dl1ekm** [thehighnibble/thehighnibble.github.io#174] [Q&A]
Just started to play around, watching YT etc. A "central" documentation of the features would really be helpful ;-)
But my question: I'm playing - also for showcase/demo - with entering programs via the switches. Cool feeling.
Is there any chance how to save (and reload) the program/memory from the webgui? I know I can run CPM and use DDT - I assume I have to enter my programs starting at 0100h in the CP-A? Any other chance to access the memory without CPM?
From another retro project (based on Atmel MCUs) I remember the feature to directly access the memory (and more) from this "hypervisor"-monitor.
Thanks at all
Marcel

**Comment by udo-munk** [comment]
Yes and No ;-) The Z80pack machines also include an emulation of a Mostek In Circuit Emulator (debugger) and options to save/load memory, which one can inspect then with the ICE. This is for the PC versions of the emulations running some sort of OS, like MacOS, Linux, Windows. The ESP32 machines don't include this and so cannot save/load memory to/from files, there is no OS, you need to use the PC versions for this.

On the ESP32 machines you need to run some OS and a debugger, which can load/save memory, but I think you figured this already. If you enter something with the front panel into the machine without that it is running an OS, you still can save memory to paper tape, which you then can save on your PC via Web GUI. But for this you must run at least some sort of monitor program, that can save on paper tape, and usually the monitors also can read in a tape. Otherwise memory content is lost when you switch the machine off.

**Comment by dl1ekm** [comment]
Thanks for the idea with the paper tape. I'll have to figure out if I can use it without having cpm running.
I think just a small memory dumper or dsave/dload would be helpful. But if I'm the only one with this suggestion - doesn't matter 😁.

Mit freundlichen Grüßen from Germany
Marcel

Am 8. Oktober 2024, um 13:25, schrieb Udo Munk ***@***.***>:

Yes and No ;-) The Z80pack machines also include an emulation of a Mostek In Circuit Emulator (debugger) and options to save/load memory, which one can inspect then with the ICE. This is for the PC versions of the emulations running some sort of OS, like MacOS, Linux, Windows. The ESP32 machines don't include this and so cannot save/load memory to/from files, there is no OS, you need to use the PC versions for this.

On the ESP32 machines you need to run some OS and a debugger, which can load/save memory, but I think you figured this already. If you enter something with the front panel into the machine without that it is running an OS, you still can save memory to paper tape, which you then can save on your PC via Web GUI. But for this you must run at least some sort of monitor program, that can save on paper tape, and usually the monitors als can read in a tape. Otherwise memory content is lost when you switch the machine off.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.￼Message ID: ***@***.***>

[ { ***@***.***": "<a href="http://schema.org">http://schema.org</a>", ***@***.***": "EmailMessage", "potentialAction": { ***@***.***": "ViewAction", "target": "<a href="https://github.com/orgs/thehighnibble/discussions/174#discussioncomment-10878909">https://github.com/orgs/thehighnibble/discussions/174#discussioncomment-10878909</a>", "url": "<a href="https://github.com/orgs/thehighnibble/discussions/174#discussioncomment-10878909">https://github.com/orgs/thehighnibble/discussions/174#discussioncomment-10878909</a>", "name": "View Discussion" }, "description": "View this Discussion on GitHub", "publisher": { ***@***.***": "Organization", "name": "GitHub", "url": "<a href="https://github.com">https://github.com</a>" } } ]

**Comment by udo-munk** [comment]
Yes, is possible, I played a bit with the paper tape without running CP/M. I do remember that we loaded the Martini animation into dazzlemation, this program can read paper tape. I also probably saved BASIC programs on paper tape. If I remember right the MEMON/80 monitor can save load to the other serial port with the paper tape. If nothing helps you would have to write litte programs that send out memory contents to the serial port and then can read it in again, not sooo difficult. Or use some such monitor someone else implemented, I have seen mentioned it, but have no pointer handy, sorry. Ages ago I messed with real paper tape at work, was enough for this life and I am not really interested nowadays ;-)

**Reply by dl1ekm** [reply]
Got working asm, load, ddt etc. and generated with SAVE a binary file of the first 256 TP memory (t1.bin) containing a running blinking led program.
DUMP t1.bin shows correct data.
With PIP PUN:=t1.bin I wrote to paper tape
With PIP t2.bin=rdr: I read back
But in t2.bin only 7 bits seems to be taken into account. Is this works as designed? Any way to store 8-bit data with CP/M tools on papertape?

**Comment by udo-munk** [comment]
Works as designed, by default the UART configuration will strip the parity bit. But this can be changed in the system configuration, so that the UART works 8bit clean. This should be documented and there are comments in the system configuration file regarding this.

**Reply by dl1ekm** [reply]
Just one more question. Tried to use the Paper Reader with XYBasic as you showed in your YT-Video.
Which version do I have to use? My "save" writes a file to (virtual) disk only.

**Comment by udo-munk** [comment]
Sounds like you tried to use the CP/M version, this saves to disk files and has no paper tape support. Use the standalone ROM version: https://github.com/udo-munk/z80pack/tree/master/imsaisim/roms
